garage/plotter/plotter.py

```diff
@@ -1,88 +1,153 @@
-    if not (process and queue):
-        init_worker()
-
-    # Needed in order to draw glfw window on the main thread
-    if ('Darwin' in platform.platform()):
-        rollout(env, policy, max_path_length=np.inf, animated=True, speedup=5)
-
-    queue.put(['update', env, policy])
+                else:
+                    if max_length:
+                        rollout(
+                            env,
+                            policy,
+                            max_path_length=max_length,
+                            animated=True,
+                            speedup=5)
+        except KeyboardInterrupt:
+            pass
+
+    def shutdown(self):
+        if self._process.is_alive():
+            self._queue.put(Message(op=Op.STOP, args=None, kwargs=None))
+            self._queue.close()
+            self._process.join()
+
+    @staticmethod
+    def set_enable(enable):
+        """Set the plot enabling according to the run_experiment function.
+        """
+        Plotter.enable = enable
+
+    @staticmethod
+    def status(plot_algo_status):
+        """Use this method as a guard for plot enabling.
+
+        This guard method considers the plot enabling defined in the
+        function run_experiment.
+
+        Parameters
+        ----------
+        plot_algo_status : boolean
+            The plot enabling defined in the algorithm class.
+
+        Returns
+        -------
+            The plot enabling of the algorithm if the run_experiment was not
+            called, and the plot enabling of the run_experiment otherwise.
+
+        """
+        if Plotter.enable is None:
+            return plot_algo_status
+        else:
+            return Plotter.enable
+
+    def init_worker(self):
+        if Plotter.enable is None or Plotter.enable:
+            self._queue = Queue()
+            if ('Darwin' in platform.platform()):
+                self._process = Thread(target=self._worker_start)
+            else:
+                self._process = Process(target=self._worker_start)
+            self._process.daemon = True
+            self._process.start()
+            atexit.register(self.shutdown)
+
+    def init_plot(self, env, policy):
+        if Plotter.enable is None or Plotter.enable:
+            if not (self._process and self._queue):
+                self.init_worker()
+
+            # Needed in order to draw glfw window on the main thread
+            if ('Darwin' in platform.platform()):
+                rollout(
+                    env,
+                    policy,
+                    max_path_length=np.inf,
+                    animated=True,
+                    speedup=5)
 
+            self._queue.put(
+                Message(op=Op.UPDATE, args=(env, policy), kwargs=None))
 
-def update_plot(policy, max_length=np.inf):
-    queue.put(['demo', policy.get_param_values(), max_length])
+    def update_plot(self, policy, max_length=np.inf):
+        if Plotter.enable is None or Plotter.enable:
```

Comment:
equivalent to `if Plotter.enable`

Reply:
err no, now i see

`if Plotter.enable is None` is equivalent to `if not Plotter.enable`

See my comment below about why this is a confusing way to do it.

Reply:
I changed it as you suggested Ryan, however, there's one tricky case when plot=True in run_experiment and plot=False in the constructor of the algorithms. Since the algorithms check their own plot flag, they're not aware of the static plot flag in the Plotter class unless the flag is overridden directly in the classes of the algorithms.
I did this in the constructors by performing this operation:
`self.plot = self.plot or Plotter.enable`
The guard in the Plotter class will take care of the cases when plot=False in the function run_experiment.

Reply:
Hmm this is still evidence then that the abstraction is leaking. The algorithm shouldn't have to check that (and new algorithm authors are unlikely to know to check it).

Perhaps the solution is to remove the `plot` parameter from either algorithm constructors or `run_experiment` altogether? Or maybe it's something more clever. Another solution is for all the Plotter functions to noop when `run_experiment` has `plot=False`.

The reason that there is a `plot` parameter in `run_experiment` is so that people can easily turn off plotting globally when they are running large jobs on headless servers. Not everyone uses run_experiment all the time, though, so moving `plot=` to *only* `run_experiment` also creates friction.

How would you design it? Put yourself in the user's shoes.